lib/log/internal/sinkcores/sentrycore/worker.go

```diff
@@ -0,0 +1,164 @@
+package sentrycore
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/getsentry/sentry-go"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
+	// flushDelay defines the brief window in which log messages are still accepted
+	// before flushing.
+	flushDelay = 500 * time.Millisecond
+	// sentryTimeout defines how much time Sentry has to send the events.
+	sentryTimeout = 5 * time.Second
+)
+
+// worker encapsulate the process of sending events to Sentry by asynchronously
+// consuming Cores, turning them into capturable, annotated errors.
+//
+// See https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/getsentry/sentry-go%24+file:%5Etransport%5C.go+defaultBufferSize&patternType=literal for more details.
+type worker struct {
+	// hub is an isolated Sentry context used to send out events to their API.
+	hub sentryHub
+	// C is the channel used to pass errors and their associated context to the go routine sending out events.
+	C chan *Core
```

Comment:
This should just be: 

```suggestion
	// C is the channel used to pass errors and their associated context to the go routine sending out events.
	C chan * errorContext
```

Then we wouldn't need to copy `Core` so frequently

Reply:
It felt simpler semantically speaking, but avoiding one clone is probably worth it, you're right. It's not like we're going to have ten errors attached to a log message.

Reply:
This sent me into a benchmarking rabbithole, and the version with copying and without mutex is actually marginally faster than the one without muxex and no copying:


No copy on write, with a mutex to protect `c.errs` 


```
BenchmarkWithSentry-10           1138370              5380 ns/op           11044 B/op        101 allocs/op
```

Before: 

```
BenchmarkWithSentry-10           1150501              5217 ns/op           10730 B/op        101 allocs/op
```

Considering those are in the end, equal (it's not like 100ns makes a difference tbh), I feel like sending cores over the channel is more aligned semantically and the code is simpler so I'd rather stick with the current implementation.

Reply:
>  I feel like sending cores over the channel is more aligned semantically with zapcore and the code is simpler so I'd rather stick with the current implementation.

😬 Is it? Could you link to example implementations that does it? Core cloning should only happen when passing to new loggers (this is just a write event), and sending the entire core makes it confusing what's happening (are we sending the logger to... log itself?)

Reply:
What's really happening is we're logging error messages, so `chan *errorContext` is the correct type

Reply:
It doesn't clone on `Write` ([only on With](https://sourcegraph.com/github.com/uber-go/zap/-/blob/zapcore/core.go?L73)), I probably over extended the reasoning in my comment, I meant it as it felt to be a simpler mental model, the unit of work are cores and that's what were sending to the other go routine :)

Reply:
Exactly - we only send things to this `chan` on `Write`, hence we shouldn't clone on `Write` either. In `With`, it is appropriate because `With` must create a new core and logger

>  the unit of work are cores

I don't think that's correct in the `zapcore` model - `zapcores` are the `io.Writer` equivalent for _logging_, where when the data is ready to render we should convert it to a concrete type to pass on to the _real_ `io.Writer` equivalent for _writing_.

Reply:
Pushed :)